### DISCUSSION TITLE
RelationshipProperty

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#11024] [Usage Questions]
### Describe the typing issue

I'm having a couple of issues with typing when trying to use a relationship (i.e. something defined as `Mapped[list[...]] = relationship()`). I'm getting an attribute with `getattr()` and therefore need to define the type I know it will be. I've tried using `orm.Relationship[object]` and `orm.RelationshipProperty[object]`, but both give me type errors.

Maybe I'm just using the wrong type, but I'm not clear how I would figure out the correct type.

### To Reproduce

```python
relationship: RelationshipProperty[object]
relationship.entity.persist_selectable.name  # error: "FromClause" has no attribute "name"  [attr-defined]

selectinload(relationship)  # error: Argument 1 to "selectinload" has incompatible type "Relationship[object]"; expected "Literal['*'] | QueryableAttribute[Any]"  [arg-type]
```

### Versions

- Python: 3.8+
- SQLAlchemy: 2.0.27
- Type checker: mypy 1.8

**Comment by CaselIT** [comment]
Hi,

The type of `SomeModel.some_relationship` is `InstrumentedAttribute[...]`. it's documented at https://docs.sqlalchemy.org/en/20/orm/internals.html#sqlalchemy.orm.InstrumentedAttribute

**Reply by Dreamsorcerer** [reply]
OK, I thought that was more generic than a relationship.

However, it appears there some typing missing there as I get an Any warning because `relationship.entity` is `Any`.

**Reply by Dreamsorcerer** [reply]
Also, it appears that it can be annotated as `InstrumentedAttribute` or `InstrumentedAttribute[A, B]`. The documentation doesn't appear to explain this, so what is A and B?

**Reply by CaselIT** [reply]
`InstrumentedAttribute` is generic on the python type, so in case of a relationship it's `InstrumentedAttribute[RelatedModel]`

**Reply by Dreamsorcerer** [reply]
Mypy tells me that it must have 2 types, which is why I'm confused (`Type application has too few types (2 expected)`). But, presumably if I want it to be generic to a relationship I'd be looking at something like `InstrumentedAttribute[type[DeclarativeBase] | type[DeclarativeBaseNoMeta]]` (if a single type were allowed atleast).

**Reply by Dreamsorcerer** [reply]
Hmm, I'm looking at the source code and it does look like it's generic over a single type, not sure what's going on with mypy...

**Reply by CaselIT** [reply]
mypy is clearly wrong here, I have no clue why it's telling you it requires 2 types, since it's definition has only one:
https://github.com/sqlalchemy/sqlalchemy/blob/9b516a9a5fc9e2a1522a5af972bedaeb8926f35b/lib/sqlalchemy/orm/attributes.py#L506

try updating mypy and verify that no other sqlalchemy stubs are installed

**Reply by Dreamsorcerer** [reply]
Hmm, so this is fine:
`foo: InstrumentedAttribute[type[DeclarativeBase]]`
But, this causes the error:
`foo = InstrumentedAttribute[type[DeclarativeBase]]`

I'm trying to do runtime type-checking with Pydantic, so I'm using it not solely as an annotation, as in the latter version.

**Reply by CaselIT** [reply]
It should not be `type[DeclarativeBase]` but just `DeclarativeBase` the generic.

I'm also not sure you want to use `DeclarativeBase` there

**Reply by Dreamsorcerer** [reply]
Well, any value I put in causes the same error:
```
aiohttp_admin/backends/sqlalchemy.py:31:21: error: Type application has too few types (2 expected)  [misc]
    _RelationshipAttr = InstrumentedAttribute[Any]
```

**Reply by Dreamsorcerer** [reply]
Runtime behaviour is definitely right, so I guess I'll open a mypy bug.

**Reply by Dreamsorcerer** [reply]
Ah, it's caused by the second TypeVar in the init method:
https://github.com/python/mypy/issues/13437

Definitely a mypy bug.
